### PR TITLE
DDF-2900 Added proximity search functionality

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
This PR adds Proximity Search functionality to Solr as well as support for it in the CQL endpoint
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@glenhein @rzwiefel 
#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
@jrnorth 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith
@pklinef
@millerw8 
#### How should this be tested? (List steps with links to updated documentation)
1.  Build 
2.  Install 
3.  Ingest a metacard
4.  Change the metacard's title to 'cat and or not the dog'
5. Use POSTMAN to issue proximity searches through the CQL endpoint and verify proper responses:

Example : 
`{"src":"ddf.distribution","start":1,"count":250,"cql":"proximity(title, 2, 'cat dog') = TRUE","sort":"modified:desc","id":"844a7d4b-ef32-488a-ad4e-43cb5c11ebea"}`

proximity(title, 2, 'cat dog') => empty result
proximity(title, 2, 'not dog') => metacard result
proximity(title, 3, 'and dog') => metacard result
proximity(title, 3, 'dog and') => empty result

6.  Repeat tests with anyText and description changed to 'cat and or not the dog'

#### What are the relevant tickets?
[DDF-2900](https://codice.atlassian.net/browse/DDF-2900)
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrFilterDelegate.java [hunk 1]
```diff
@@ -755,6 +754,25 @@ public SolrQuery xpathIsFuzzy(String xpath, String literal) {
...
+    }
+
+    @Override
+    public SolrQuery propertyIsNotInProximityTo(String propertyName, Integer distance, String searchTerms) {
+        return getProximityQuery(propertyName, distance, searchTerms, false);
+    }
+
+    private SolrQuery getProximityQuery(String propertyName, Integer distance, String searchTerms, boolean is) {
+        String not = is ? "" : "!";
+        String searchPhrase = QUOTE + searchTerms + QUOTE + " ~" + distance;
```
Comment (glenhein) [inline catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrFilterDelegate.java]: Can 'searchTerms' contain a quote character? Do we need to make sure 'distance' is always >0?

File: catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrFilterBuilderTest.java [hunk 2]
```diff
@@ -47,8 +51,19 @@ public Filter dateGreaterThanOrEqual(String attribute, Date literal) {
...
     public Filter dateIsDuring(String attribute, Date lowerBoundary, Date upperBoundary) {
-        return factory.during(factory.property(attribute),
-                factory.literal(makePeriod(lowerBoundary, upperBoundary)));
+        return factory.during(factory.property(attribute), factory.literal(makePeriod(lowerBoundary,
+                upperBoundary)));
+    }
+
+    public Filter proximity(String attribute, int distance, String searchTerms) {
+        FunctionImpl function = new FunctionImpl();
+        function.setName("proximity");
```
Comment (glenhein) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrFilterBuilderTest.java]: see earlier comment about making 'proximity' a constant

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java [hunk 3]
```diff
@@ -129,6 +139,21 @@ public void init(ServletConfig config) throws ServletException {
                 false);
     }
 
+    private void registerProximityFactory() {
```
Comment (glenhein) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java]: Should this method make sure that the proximity function factory wasn't already registered?

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java [hunk 4]
```diff
@@ -949,7 +976,10 @@ private ExpressionValues getExpressions(Expression expression1, Expression expre
         String propertyName;
         Object literal;
 
-        if (expression1 instanceof PropertyName && expression2 instanceof Literal) {
+        if (expression1 instanceof Function && expression2 instanceof Literal) {
+            return new ExpressionValues((ProximityArguments) expression1.accept(this, delegate),
```
Comment (millerw8) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: How do we know if it is a function we should be looking for ProximityArguments?  Does this mess up future extensions?

File: catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/ProximityFunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.Test;
+import org.opengis.filter.expression.Expression;
+
+public class ProximityFunctionTest {
+
+    @Test(expected = NullPointerException.class)
+    public void testVerifyFuzzyFunctionCannotBeCalledWithNull() {
```
Comment (millerw8) [inline catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/ProximityFunctionTest.java]: Are these functions all copy-paste errors in their names?

File: catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    public static final FunctionName NAME = functionName("proximity",
+            "result:Boolean",
+            "property:String",
+            "distance:Integer",
+            "text:String:2,");
+
+    public ProximityFunction(List<Expression> parameters, Literal fallback) {
+
+        if (parameters == null) {
```
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java]: Changed

File: catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        this.setName("proximity");
+        this.setParameters(parameters);
+        this.setFallbackValue(fallback);
+        this.functionName = NAME;
+    }
+
+    @Override
+    public List<Expression> getParameters() {
```
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java]: Fixed

File: catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        return super.getParameters();
+    }
+
+    @Override
+    public Object evaluate(Object object) {
```
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/ProximityFunction.java]: Fixed

File: catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/GeoToolsFunctionFactoryTest.java [hunk 9]
```diff
@@ -25,48 +27,57 @@
...
 
     @Test
     public void testGetFunctionNames() {
         List<FunctionName> funcs = toTest.getFunctionNames();
-        assertEquals(1, funcs.size());
-        assertEquals("fuzzy",
-                funcs.get(0)
-                        .getName());
+        assertThat(funcs, hasSize(2));
+        assertThat(funcs.get(0).getName(), is("fuzzy"));
```
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/GeoToolsFunctionFactoryTest.java]: Since the GeoToolsFunctionFactory adds the functions to an ArrayList (sequential) in the getFunctionNames call, we can assume that the retrieval order will be the same.
https://github.com/bdeining/ddf/blob/e12133920525e9610a0420da2c2a2d96fec2ac92/catalog/core/catalog-core-commons/src/main/java/ddf/catalog/impl/filter/GeoToolsFunctionFactory.java#L38

File: catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/GeoToolsFunctionFactoryTest.java [hunk 10]
```diff
@@ -25,48 +27,57 @@
...
+        assertThat(result.getName(), is("fuzzy"));
+    }
+
+    @Test
+    public void testProximityFunction() {
+        List<Expression> expr = new ArrayList<>();
+        expr.add(Expression.NIL);
+        expr.add(Expression.NIL);
+        expr.add(Expression.NIL);
+        Function result = toTest.function("proximity", expr, null);
```
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/GeoToolsFunctionFactoryTest.java]: It's defined in ProximityFunction; fixed to used constant instead

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java [hunk 11]
```diff
@@ -134,6 +135,22 @@ public Object visit(Function expression, Object delegate) {
             throw new UnsupportedOperationException(
                     Function.class.getSimpleName() + " must not be null.");
         }
+
+        if (expression.getName().equals(ProximityFunction.NAME.getName())) {
+            List<Expression> parameters = expression.getParameters();
+            if (parameters.size() == 3) {
```
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: The number of parameters of the function should only be 3; the query should fail otherwise.  Updated this to be a constant in ProximityFunction

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java [hunk 12]
```diff
@@ -134,6 +135,22 @@ public Object visit(Function expression, Object delegate) {
...
+
+                String propertyName = parameters.remove(0)
+                        .toString();
+                int distance = Integer.parseInt(parameters.remove(0)
```
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: Added a catch

File: catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrFilterDelegate.java (hunk 1 above)
Comment (bdeining) [inline catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrFilterDelegate.java]: Looking at this some more, searchTerms should call the escapeSpecialChars method (which handles quote characters).  Added a check for distance

File: catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrFilterBuilderTest.java (hunk 2 above)
Comment (bdeining) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrFilterBuilderTest.java]: fixed

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java (hunk 4 above)
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: This currently doesn't mess up future extensions because this is the first function to be added to `getExpressions`, but it should be more granular.  I've updated this conditional to specifically check for ProximityFunctions

File: catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/ProximityFunctionTest.java (hunk 5 above)
Comment (bdeining) [inline catalog/core/catalog-core-commons/src/test/java/ddf/catalog/impl/filter/ProximityFunctionTest.java]: Fixed

**Comment by bdeining** [comment]
@millerw8 adding the ProximityFunction to the GeoTools FactoryFinder allows the parsing of CQL into a filter (under the GeoTools hood) that will in turn get propagated down to the SolrFilterDelegate

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java (hunk 3 above)
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java]: I'm not sure that that is necessary.  This function is only called during the `init()` of the Catalog UI; the FunctionFactory should be registered with GeoTools on initialization

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/filter/FilterDelegate.java [hunk 13]
```diff
@@ -1239,4 +1239,14 @@ public T relative(String propertyName, long duration) {
                 "relative(String,long) not supported by org.opengis.filter.Filter Delegate.");
     }
 
+    public T propertyIsInProximityTo(String propertyName, Integer distance, String searchTerm) {
```
Comment (jrnorth) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/filter/FilterDelegate.java]: Javadoc?

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java [hunk 14]
```diff
@@ -134,6 +135,27 @@ public Object visit(Function expression, Object delegate) {
             throw new UnsupportedOperationException(
                     Function.class.getSimpleName() + " must not be null.");
         }
+
+        if (expression.getName().equals(ProximityFunction.NAME.getName())) {
+            List<Expression> parameters = expression.getParameters();
+            if (parameters.size() == ProximityFunction.NUM_PARAMETERS) {
+
+                String propertyName = parameters.remove(0)
```
Comment (jrnorth) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: I'm not familiar with the `ExpressionVisitor` API - is it okay to remove parameters from the `Function`?

**Comment by millerw8** [comment]
Trying to run this, I got a successful result but saw this in the logs:

`17:42:14,025 | INFO  | p1095489863-1013 | g.cache.solr.impl.ValidationQueryFactory   57 | re-standardframework | This attribute filter is not supported by ValidationQueryDelegate.
ddf.catalog.source.UnsupportedQueryException: propertyIsInProximityTo(String,Integer,String) not supported by org.opengis.filter.Filter Delegate.
        at ddf.catalog.filter.proxy.adapter.GeotoolsFilterAdapterImpl.adapt(GeotoolsFilterAdapterImpl.java:116)
        at Proxy6fd2aff8_01f9_49b9_91fb_37eb8431457e.adapt(Unknown Source)
        at ddf.catalog.cache.solr.impl.ValidationQueryFactory.getQueryRequestWithValidationFilter(ValidationQueryFactory.java:51)
        at ddf.catalog.cache.solr.impl.CachingFederationStrategy.sourceFederate(CachingFederationStrategy.java:266)
        at ddf.catalog.cache.solr.impl.CachingFederationStrategy.federate(CachingFederationStrategy.java:199)
        at ddf.catalog.impl.operations.QueryOperations.doQuery(QueryOperations.java:207)
        at ddf.catalog.impl.operations.QueryOperations.query(QueryOperations.java:158)
        at ddf.catalog.impl.operations.QueryOperations.query(QueryOperations.java:122)
        at ddf.catalog.impl.operations.QueryOperations.query(QueryOperations.java:116)
        at ddf.catalog.impl.CatalogFrameworkImpl.query(CatalogFrameworkImpl.java:282)
        at Proxy7b42f9b8_7b5d_486b_928e_c3884b2da081.query(Unknown Source)
        at org.codice.ddf.catalog.ui.query.QueryApplication.executeCqlQuery(QueryApplication.java:102)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at org.codice.ddf.catalog.ui.query.QueryApplication.lambda$init$0(QueryApplication.java:70)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at spark.RouteImpl$1.handle(RouteImpl.java:61)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at spark.http.matching.Routes.execute(Routes.java:61)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at org.codice.ddf.catalog.ui.SparkServlet.service(SparkServlet.java:181)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)[10:javax.servlet-api:3.1.0]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)[73:org.eclipse.jetty.servlet:9.2.19.v20160908]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)[73:org.eclipse.jetty.servlet:9.2.19.v20160908]
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:119)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.platform.filter.clientinfo.ClientInfoFilter.doFilter(ClientInfoFilter.java:75)
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:112)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.platform.response.filter.ResponseFilter.doFilter(ResponseFilter.java:81)
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:112)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.security.filter.authorization.AuthorizationFilter.doFilter(AuthorizationFilter.java:105)
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:112)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.security.filter.login.LoginFilter.lambda$null$0(LoginFilter.java:279)[268:security-filter-login:2.11.0.SNAPSHOT]
        at java.security.AccessController.doPrivileged(Native Method)[:1.8.0_102]
        at javax.security.auth.Subject.doAs(Subject.java:422)[:1.8.0_102]
        at org.codice.ddf.security.filter.login.LoginFilter.lambda$doFilter$1(LoginFilter.java:291)[268:security-filter-login:2.11.0.SNAPSHOT]
        at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)[197:org.apache.shiro.core:1.3.2]
        at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)[197:org.apache.shiro.core:1.3.2]
        at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)[197:org.apache.shiro.core:1.3.2]
        at org.codice.ddf.security.filter.login.LoginFilter.doFilter(LoginFilter.java:277)[268:security-filter-login:2.11.0.SNAPSHOT]
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:112)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.security.filter.websso.WebSSOFilter.handleRequest(WebSSOFilter.java:234)[269:security-filter-web-sso:2.11.0.SNAPSHOT]
        at org.codice.ddf.security.filter.websso.WebSSOFilter.doFilter(WebSSOFilter.java:129)[269:security-filter-web-sso:2.11.0.SNAPSHOT]
        at org.codice.ddf.platform.filter.delegate.ProxyFilterChain.doFilter(ProxyFilterChain.java:112)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.codice.ddf.platform.filter.delegate.DelegateServletFilter.doFilter(DelegateServletFilter.java:105)[107:platform-filter-delegate:2.11.0.SNAPSHOT]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)[73:org.eclipse.jetty.servlet:9.2.19.v20160908]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)[73:org.eclipse.jetty.servlet:9.2.19.v20160908]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71)[90:org.ops4j.pax.web.pax-web-jetty:4.3.0]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)[71:org.eclipse.jetty.security:9.2.19.v20160908]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:287)[90:org.ops4j.pax.web.pax-web-jetty:4.3.0]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)[73:org.eclipse.jetty.servlet:9.2.19.v20160908]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)[90:org.ops4j.pax.web.pax-web-jetty:4.3.0]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.Server.handle(Server.java:499)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)[72:org.eclipse.jetty.server:9.2.19.v20160908]
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)[64:org.eclipse.jetty.io:9.2.19.v20160908]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)[75:org.eclipse.jetty.util:9.2.19.v20160908]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)[75:org.eclipse.jetty.util:9.2.19.v20160908]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_102]
Caused by: java.lang.UnsupportedOperationException: propertyIsInProximityTo(String,Integer,String) not supported by org.opengis.filter.Filter Delegate.
        at ddf.catalog.filter.FilterDelegate.propertyIsInProximityTo(FilterDelegate.java:1243)[300:catalog-core-api:2.11.0.SNAPSHOT]
        at ddf.catalog.filter.proxy.adapter.GeotoolsFilterAdapterImpl.visit(GeotoolsFilterAdapterImpl.java:311)
        at org.geotools.filter.IsEqualsToImpl.accept(IsEqualsToImpl.java:162)[198:org.codice.thirdparty.geotools-suite:8.4.0.2]
        at ddf.catalog.operation.impl.QueryImpl.accept(QueryImpl.java:87)[421:catalog-ui-search:2.11.0.SNAPSHOT]
        at ddf.catalog.filter.proxy.adapter.GeotoolsFilterAdapterImpl.adapt(GeotoolsFilterAdapterImpl.java:113)
        ... 60 more

`

**Comment by millerw8** [comment]
So I see this works for the internal UI endpoint but not CSW endpoint and a CQL constraint.

Is this because internal UI does `filter = ECQL.toFilter(cql)` and CSW does `filter = ECQL.toFilter(cql)` ?

Does it work with an OGC Filter?

I was hoping this query would work on the CSW Endpoint:

```
<?xml version="1.0" ?> 
<GetRecords xmlns="http://www.opengis.net/cat/csw/2.0.2" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
            xmlns:gml="http://www.opengis.net/gml" service="CSW" version="2.0.2" maxRecords="9" startPosition="1" resultType="results" 
            outputFormat="application/xml" outputSchema="http://www.opengis.net/cat/csw/2.0.2" 
            xsi:schemaLocation="http://www.opengis.net/cat/csw/2.0.2 ../../../csw/2.0.2/CSW-discovery.xsd"> 

  <Query typeNames="Record"> 
    <ElementSetName>summary</ElementSetName> 
    <Constraint version="1.1.0"> 
      <ogc:CqlText>proximity(title, 10, 'new image') = TRUE</ogc:CqlText>
    </Constraint> 
    <ogc:SortBy>
      <ogc:SortProperty>
       <ogc:PropertyName>title</ogc:PropertyName>
       <ogc:SortOrder>ASC</ogc:SortOrder>
      </ogc:SortProperty>
    </ogc:SortBy>
  </Query> 
</GetRecords>
```

**Comment by bdeining** [comment]
I believe the CswQueryFactory would need to be updated to use `ECQL` vs `CQL`
https://github.com/codice/ddf/blob/master/catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswQueryFactory.java#L170
I'm curious what kind of adverse effects (if any) would arise from making the switch

**Comment by millerw8** [comment]
So, if I am seeing this correctly, adjacent words have a cost of 1, as well as words that have a word between them, and reversing the order has a cost of 2.  Is that correct?

with  text = "my small home" these are the min distances: (?)
small home 1
my home 1
home small 2
home my 3

Is that the expected behavior?  I guess it is just how solr works, but doesn't feel intuitive.  I assume we will restrict the near parameter to 1?

**Comment by bdeining** [comment]
Correct, in general, reversing the order costs an additional 2 distance because something like "home small" is technically 2 changes away from "small home" (internally the positions of small and home must be swapped).  Thus, this is expected behavior.  We shouldn't restrict the near parameter, however, as that would prevent the ability to do a proximity search like the following : 
"my big small home" --> "my home" 2

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/filter/FilterDelegate.java (hunk 13 above)
Comment (bdeining) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/filter/FilterDelegate.java]: Added javadoc

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java (hunk 14 above)
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: The API doesn't specifically discuss the safety of removing the parameters here.  I changed this code to create a new list just to be safe.

**Comment by bdeining** [comment]
@millerw8 fixed the ValidationQueryDelegate to prevent those warnings

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java [hunk 15]
```diff
@@ -949,7 +981,10 @@ private ExpressionValues getExpressions(Expression expression1, Expression expre
         String propertyName;
         Object literal;
 
-        if (expression1 instanceof PropertyName && expression2 instanceof Literal) {
+        if (expression1 instanceof Function && ((Function) expression1).getFunctionName().getName().equals(ProximityFunction.NAME.getName()) && expression2 instanceof Literal) {
```
Comment (pklinef) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: I would suggest reversing the equals to avoid any potential NPE.

File: catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java [hunk 16]
```diff
@@ -95,25 +96,25 @@
             new DefaultPeriodDuration(DAY_IN_MILLISECONDS);
 
     @Test
-    public void includeFilter() {
+    public void testIncludeFilter() {
```
Comment (pklinef) [inline catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java]: Why did you add test to the name of every test method?  That is not required with the annotation.

File: catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java [hunk 17]
```diff
@@ -3322,6 +3352,30 @@ public void testPropertyIsEqualTo() throws Exception {
...
+
+        list.add(metacard);
+        list.add(metacard1);
+
+        create(list);
+
+        queryAndVerifyCount(1, filterBuilder.proximity(Core.TITLE, 3, "Mary ham"));
+        queryAndVerifyCount(0, filterBuilder.proximity(Core.TITLE, 2, "Mary ham"));
+        queryAndVerifyCount(2, filterBuilder.proximity(Core.TITLE, 2, "Mary little"));
+        queryAndVerifyCount(0, filterBuilder.proximity(Core.TITLE, 1, "Mary little"));
```
Comment (pklinef) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java]: Does proximity work with anyText queries?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java [hunk 18]
```diff
@@ -112,6 +120,8 @@ public void setFilterMappingPattern(String filterMappingPattern) {
 
     @Override
     public void init(ServletConfig config) throws ServletException {
+        registerProximityFactory();
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java]: Any way we can move this to blueprint instead of putting it into the `SparkSevlet`?  There is a PR up on Spark to get the SparkServlet merged into their master so this file might be removed in the future.

File: catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java (hunk 16 above)
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java]: Our Unit Test guidelines state that the test prefix is optional.  The majority of the unit tests within this package are prefixed with test, so this was changed for continuity and uniformity.
https://codice.atlassian.net/wiki/display/DDF/Unit+Testing

File: catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java (hunk 17 above)
Comment (bdeining) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java]: Verified that proximity works with anyText queries

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java (hunk 18 above)
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java]: Moved to `EndpointUtil`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java (hunk 3 above)
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/SparkServlet.java]: Renamed

File: catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java (hunk 15 above)
Comment (bdeining) [inline catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java]: Switched

File: catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java (hunk 16 above)
Comment (pklinef) [inline catalog/core/catalog-core-impl/filter-proxy/src/test/java/ddf/catalog/filter/proxy/adapter/test/FilterAdapterTest.java]: I figured that was the case but wanted to confirm.  Thanks for making it consistent.

File: catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java (hunk 17 above)
Comment (pklinef) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java]: Should we add a test to make sure that does not regress?

Comment (bdeining) [inline catalog/core/catalog-core-solr/src/test/java/ddf/catalog/source/solr/SolrProviderTest.java]: Added a unit test for anyText queries

**Comment by bdeining** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by millerw8** [comment]
build now

**Comment by millerw8** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by bdeining** [comment]
build now
